pkg/nodejs/pnpm/parse.go

```diff
@@ -100,10 +100,73 @@ func isIndirectLib(name string, directDeps map[string]string) bool {
 	return !ok
 }
 
-func getPackageNameAndVersion(pkg string) (string, string) {
-	idx := strings.LastIndex(pkg, "/")
-	name := pkg[1:idx]
-	version := pkg[idx+1:]
+func getPackageNameAndVersion(pkg string, lockFileVersion int8) (string, string) {
+	if lockFileVersion < 6 {
+		return getPackageNameAndVersionV5(pkg)
+	}
+	return getPackageNameAndVersionV6(pkg)
+}
+
+// 2 package formats are possible:
+// relative path: `/<pkg_name>/<pkg_version>` e.g. /foo/1.0.0
+// registry: `<registry_url>/<pkg_name>/<pkg_version>` e.g. registry.node-modules.io/foo/1.0.0
+// https://github.com/pnpm/spec/blob/master/lockfile/5.2.md#packages
+//
+// `pkg_name` has 2 formats:
+// with slash - @<author>/<name>. e.g. `/@babel/generator/7.21.9`
+// without slash - <name>. e.g. `/lodash/4.17.10`
+//
+// `pkg_version` can contain peer deps. <pkg_name>/<pkg_version>_<peer_deps>
+// e.g. /@babel/helper-compilation-targets/7.21.5_@babel+core@7.21.8
+func getPackageNameAndVersionV5(pkg string) (string, string) {
+	s := strings.Split(pkg, "/")
+	// take name as last element before version
+	name := s[len(s)-2]
+	// if previous element start from `@` => this is name with slash
+	if strings.HasPrefix(s[len(s)-3], "@") {
+		name = strings.Join(s[len(s)-3:len(s)-1], "/")
+	}
+
+	version := s[len(s)-1]
+	// trim peer deps
+	if strings.Contains(version, "_") {
+		version = version[:strings.Index(version, "_")]
+	}
+	return name, version
+}
+
+//	2 package formats are possible:
+//
+// relative path: `/<pkg_name>@<pkg_version>` e.g. /foo@1.0.0
+// registry: `<registry_url>/<pkg_name>@<pkg_version>` e.g. registry.node-modules.io/foo@1.0.0
+// https://github.com/pnpm/pnpm/pull/5810
+// https://github.com/pnpm/spec/issues/4#issuecomment-1558891433
+//
+// `pkg_name` has 2 formats:
+// with slash - @<author>@<name>. e.g. `/@babel/generator@7.21.9`
+// without slash - <name>. e.g. `/lodash@4.17.10`
+//
+// `pkg_version` can contain peer deps. <pkg_name>@<pkg_version>(<peer_deps>)
```

Comment:
I think we can simplify the implementation.
https://go.dev/play/p/KyThjdKYGRi

```suggestion
func parsePackage(pkg, versionSep string) (string, string) {
	// Skip registry
	// e.g.
	//    - "registry.npmjs.org/lodash/4.17.10" => "lodash/4.17.10"
	//    - "registry.npmjs.org/@babel/generator/7.21.9" => "@babel/generator/7.21.9"
	//    - "/lodash/4.17.10" => "lodash/4.17.10"
	_, pkg, _ = strings.Cut(pkg, "/")

	// Parse namespace(?)
	// e.g.
	//    - v5:  "@babel/generator/7.21.9" => {"babel", "generator/7.21.9"}
	//    - v6+: "@babel/helper-annotate-as-pure@7.18.6" => "{"babel", "helper-annotate-as-pure@7.18.6"}
	var namespace string
	if strings.HasPrefix(pkg, "@") {
		namespace, pkg, _ = strings.Cut(pkg, "/")
	}

	// Parse package name
	// e.g.
	//    - v5:  "generator/7.21.9" => {"generator", "7.21.9"}
	//    - v6+: "helper-annotate-as-pure@7.18.6" => {"helper-annotate-as-pure", "7.18.6"}
	var name, version string
	name, version, _ = strings.Cut(pkg, versionSep)

	if namespace != "" {
		name = fmt.Sprintf("%s/%s", namespace, name)
	}

	// Trim peer deps
	// e.g.
	//    - v5:  "7.21.5_@babel+core@7.21.8" => "7.21.5"
	//    - v6+: "7.21.5(@babel/core@7.20.7)" => "7.21.5"
	if idx := strings.IndexAny(version, "_("); idx != -1 {
		version = version[:idx]
	}

	return name, version
}
```

Reply:
this way doesn't work with repository name with additional `/`. I mean `npm-proxy.fury.io/dmitriytest/lodash/5.0.0`.
I added test for that - https://github.com/aquasecurity/go-dep-parser/pull/220/files#diff-828b2b43cb2b887728250ad289067f4874a1e6d0a52fb6d339f1020c0f4c28ecR105-R111

Reply:
I didn't see such a case in the tests, so I didn't expect the registry also contains a slash.

Reply:
I read through pnpm logic. It doesn't allow more slashes in registries.
https://github.com/pnpm/dependency-path/blob/02815d16d05d513a5103feaeafdf88716cb72bce/src/index.ts#L61-L86

Reply:
Looks like they adopt the same approach as mine.

Reply:
This is probably a new one.
https://github.com/pnpm/pnpm/blob/ce61f8d3c29eee46cee38d56ced45aea8a439a53/lockfile/lockfile-file/src/experiments/inlineSpecifiersLockfileConverters.ts#L162-L167

Reply:
My conclusion so far is below.

- registries don't allow `/`.
    - We don't need the test case like `npm-proxy.fury.io/dmitriytest/lodash/5.0.0`
- Our logic is missing file support.
    - https://github.com/pnpm/pnpm/blob/ce61f8d3c29eee46cee38d56ced45aea8a439a53/lockfile/lockfile-file/test/lockfileV6Converters.test.ts#L36
    - need semver check?
        - https://github.com/pnpm/pnpm/blob/ce61f8d3c29eee46cee38d56ced45aea8a439a53/packages/dependency-path/src/index.ts#LL148C6-L148C6

Reply:
> We don't need the test case like npm-proxy.fury.io/dmitriytest/lodash/5.0.0

I used you logic.

> Our logic is missing file support.
need semver check? 

Made in https://github.com/aquasecurity/go-dep-parser/pull/220/commits/6f9e3f0612b8ca5e2031a2e249d626bdd82f7c39